File: prep/implementation/iterations/iteration_cpp/my_priority_queue.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

template <typename T> class priority_queue {
private:
  vector<T> arr;

  int parent(int p) { return p / 2; }

  int left(int p) { return 2 * p + 1; }

  int right(int p) { return 2 * p + 2; }

  void swap(int p, int q) {
    T t = this->arr[p];
    this->arr[p] = this->arr[q];
    this->arr[q] = t;
  }

  void sift_up(int p) {
    while (p != 0 && (this->arr[parent(p)] < this->arr[p])) {
      swap(p, parent(p));
      p = parent(p);
    }
  }

  void sift_down(int p) {
    int len = this->arr.size();

    int max_idx = p;

    int left_idx = left(p);
    if (left_idx<len &&this->arr[left_idx]> this->arr[max_idx]) {
      max_idx = left_idx;
    }

    int right_idx = right(p);
    if (right_idx < len && (this->arr[right_idx] > this->arr[max_idx])) {
      max_idx = right_idx;
    }

    if (max_idx != p) {
      swap(max_idx, p);
      sift_down(max_idx);
    }
  }

public:
  void add(T elem) {
    this->arr.push_back(elem);
    sift_up(this->arr.size() - 1);
  }

  bool empty() { return this->arr.empty(); }

  T poll() {
    if (empty()) {
      throw "empty priority queue";
    }

    swap(0, this->arr.size() - 1);
    T val = this->arr.back();
    this->arr.pop_back();
    sift_down(0);

    return val;
  }
};

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int n, d;
  cin >> n;

  priority_queue<int> max_heap;
  for (int i = 0; i < n; i++) {
    cin >> d;
    max_heap.add(d);
  }

  cout << max_heap.empty() << "\n";

  while (!max_heap.empty()) {
    cout << max_heap.poll() << " ";
  }
  cout << "\n";

  cout << max_heap.empty() << "\n";

  return 0;
}

File: prep/implementation/practice_iterations/main_1/MyPriorityQueue.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class MyPriorityQueue<T> {
  private List<T> arr;
  private Comparator<T> comp;

  public MyPriorityQueue(Comparator<T> comp) {
    this.arr = new ArrayList<>();
    this.comp = comp;
  }

  private int left(int p) { return 2 * p + 1; }

  private int right(int p) { return 2 * p + 2; }

  private int parent(int p) { return p / 2; }

  private void swap(int x, int y) {
    T t = this.arr.get(x);
    this.arr.set(x, this.arr.get(y));
    this.arr.set(y, t);
  }

  private void siftUp(int p) {
    while (p >= 0 &&
           comp.compare(this.arr.get(p), this.arr.get(parent(p))) > 0) {
      swap(p, parent(p));
      p = parent(p);
    }
  }

  private void siftDown(int p) {
    int maxIdx = p;

    int leftIdx = left(p);
    if (leftIdx < size() &&
        comp.compare(this.arr.get(leftIdx), this.arr.get(maxIdx)) > 0) {
      maxIdx = leftIdx;
    }

    int rightIdx = right(p);
    if (rightIdx < size() &&
        comp.compare(this.arr.get(rightIdx), this.arr.get(maxIdx)) > 0) {
      maxIdx = rightIdx;
    }

    if (maxIdx != p) {
      swap(maxIdx, p);
      siftDown(maxIdx);
    }
  }

  // O(logn)
  public void add(T elem) {
    this.arr.add(elem);
    siftUp(this.arr.size() - 1);
  }

  // O(logn)
  public T poll() {
    if (isEmpty()) {
      throw new IllegalStateException("empty priority queue");
    }

    T val = this.arr.get(0);
    swap(0, this.arr.size() - 1);
    this.arr.remove(this.arr.size() - 1);
    siftDown(0);

    return val;
  }

  // O(1)
  public boolean isEmpty() { return this.arr.isEmpty(); }

  // O(1)
  public int size() { return this.arr.size(); }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      MyPriorityQueue<Integer> maxHeap =
          new MyPriorityQueue<>((x, y) -> Integer.compare(x, y));
      MyPriorityQueue<Integer> minHeap =
          new MyPriorityQueue<>((x, y) -> Integer.compare(y, x));

      int n = in.nextInt();
      for (int i = 0; i < n; i++) {
        int elem = in.nextInt();
        maxHeap.add(elem);
        minHeap.add(elem);
      }

      System.out.println(maxHeap.isEmpty());
      while (!maxHeap.isEmpty()) {
        System.out.printf("%d ", maxHeap.poll());
      }
      System.out.println();
      System.out.println(maxHeap.isEmpty());

      System.out.println(minHeap.isEmpty());
      while (!minHeap.isEmpty()) {
        System.out.printf("%d ", minHeap.poll());
      }
      System.out.println();
      System.out.println(minHeap.isEmpty());
    }
  }
}